src/MarginCalculator.php

```diff
@@ -0,0 +1,141 @@
+     * Candidates, then this method should register (N^2) - N Candidates.
+     *
+     * @todo #7 Decide whether this method should be a part of the MarginCalculator class or the Agenda class.
+     */
+    protected function initializeRegistry(Agenda $agenda) : MarginRegistry
+    {
+        $registry = new MarginRegistry();
+        foreach ($agenda->getCandidates() as $outerCandidate) {
+            foreach ($agenda->getCandidates() as $innerCandidate) {
+                /**
+                 * It only makes sense to keep track of the difference of public support between
+                 * two DIFFERENT candidates. It doesn't make sense to keep track of the difference between
+                 * public support of a candidate and themself.
+                 */
+                if ($outerCandidate->getId() != $innerCandidate->getId()) {
+                    $margin = new Margin($outerCandidate, $innerCandidate, 0);
+                }
+            }
+        }
+        return $registry;
+    }
+    /**
+     * Return an associative array that maps Candidates' ids to an integer. The integer represents the rank of the
+     * Candidate within the Ballot. A lower index indicates higher preference. An index of zero is the highest rank.
+     * Since a Ballot can contain ties, multiple Candidate ids can map to the same integer rank.
+     */
+    protected function getCandidateIdToRankMap(Ballot $ballot) : array
+    {
+        $candidateIdToRank = array();
+        /**
+         * @todo #7 build map of candidate id to rank
+         */
+        foreach ($ballot as $rank => $candidateList) {
+            foreach ($candidateList as $candidate) {
+                $candidateId = $candidate->getId();
+                if (empty($candidateId)) {
+                    throw new \InvalidArgumentException("Candidates must have a non-empty Id");
+                } else {
+                    if (isset($candidateIdToRank[$candidateId])) {
+                        throw new \InvalidArgumentException(
+                            "A Ballot cannot contain a candidate more than once."
+                            . " Offending Ballot: " . $ballot
+                            . " Offending Candidate: " . $candidate
+                        );
+                    } else {
+                        $candidateIdToRank[$candidateId] = $rank;
+                    }
+                }
+            }
+        }
+        return $candidateIdToRank;
+    }
+
+    /**
+     * Uses the parameterized candidates and the $candidateIdToRank map to determine which candidate ranks higher.
+     * Subsequently adds the $ballotCount to the count already associated with the appropriate margin in the
+     * Registry.
+     *
+     * @param Candidate $outerCandidate
+     * @param Candidate $innerCandidate
+     * @param array $candidateIdToRank
+     * @param MarginRegistry $registry
+     * @param int $ballotCount
+     */
+    protected function updatePairInRegistry(
```

Comment:
This method is too difficult to test. Do you have any suggestions on how to break it up into smaller methods?